numpy/core/shape_base.py

```diff
@@ -275,3 +278,62 @@ def hstack(tup):
         return _nx.concatenate(arrs, 0)
     else:
         return _nx.concatenate(arrs, 1)
+
+def stack(arrays, axis=0):
+    """
+    Join a sequence of arrays along a new axis.
+
+    .. versionadded:: 1.10.0
+
+    Parameters
+    ----------
+    arrays : sequence of ndarrays
+        Each array must have the same shape.
+    axis : int, optional
+        The axis along which the arrays will be stacked.
+
+    Returns
+    -------
+    stacked : ndarray
+        The stacked array has one more dimension than the input arrays.
+
+    See Also
+    --------
+    concatenate : Join a sequence of arrays along an existing axis.
+    split : Split array into a list of multiple sub-arrays of equal size.
+
+    Examples
+    --------
+    >>> arrays = [np.random.randn(3, 4) for _ in range(10)]
+    >>> np.stack(arrays, axis=0).shape
+    (10, 3, 4)
+
+    >>> np.stack(arrays, axis=1).shape
+    (3, 10, 4)
+
+    >>> np.stack(arrays, axis=2).shape
+    (3, 4, 10)
+
+    >>> a = np.array([1, 2, 3])
+    >>> b = np.array([2, 3, 4])
+    >>> np.stack((a, b))
+    array([[1, 2, 3],
+           [2, 3, 4]])
+
+    >>> np.stack((a, b), axis=-1)
+    array([[1, 2],
+           [2, 3],
+           [3, 4]])
+
+    """
+    arrays = [_nx.asarray(arr) for arr in arrays]
```

Comment:
@mhvk I'm all for supporting ndarray subclasses when feasible, but the options we've come up with so far would either entail worse error reporting for non-compliant subclasses (using `asanyarray`) or additions to the function signature (adding `subok=False`) with no actual benefits, given the limitations of `concatenate`.

Are there any non-trivial ndarray subclasses that *would* benefit from these changes? I suppose we could argue that it is best to keep the door open to improvements in the future.